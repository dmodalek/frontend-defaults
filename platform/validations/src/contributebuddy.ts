import { ContributebuddyAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import {
	ValidationResult,
	ValidationSeverityLevel,
	validateInstallation,
} from '@namics/frontend-defaults-platform-core';

export const contributebuddyValidation = async (
	cwd: string,
	analytics: ContributebuddyAnalyzerResult
): Promise<ValidationResult[]> => {
	if (analytics.contributebuddy) {
		// contributebuddy is used in the project, validation below
		let validationResults: ValidationResult[] = [];
		validationResults = validationResults.concat(
			validateInstallation({
				installation: analytics.contributebuddyInstallation,
				name: 'Contributebuddy',
				source: 'contributebuddyValidation',
			})
		);

		if (!analytics.contributebuddyConfigFolder) {
			validationResults.push({
				level: ValidationSeverityLevel.error,
				message: 'No .contributebuddy folder found, make sure you set-up correctly',
				source: 'contributebuddyValidation',
			});
		}

		return validationResults;
	}

	return [
		{
			level: ValidationSeverityLevel.info,
			message: 'We recommend to use contribute-buddy to obtain maintainability',
			source: 'contributebuddyValidation',
		},
	];
};
