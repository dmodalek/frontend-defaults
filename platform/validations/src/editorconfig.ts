import { EditorconfigAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import { ValidationResult, ValidationSeverityLevel } from '@namics/frontend-defaults-platform-core';

export const editorconfigValidation = async (cwd: string, analytics: EditorconfigAnalyzerResult): Promise<ValidationResult[]> => {
	if (analytics.editorConfig) {
		// TODO: editorconfig is available, so we need to validate all things
		return [];
	}

	return [{
		level: ValidationSeverityLevel.error,
		message: 'Default namics editorconfig file is required in every project',
		source: 'editorconfigValidation'
	}];
}