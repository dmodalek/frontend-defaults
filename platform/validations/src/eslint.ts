import { ESLintAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import { ValidationResult, ValidationSeverityLevel, validateInstallation } from '@namics/frontend-defaults-platform-core';

/**
 * TODO: Should we check if a project is a TS project in advance? So we wouldn't need to run this validation atm.
 * I think we'll need to update this as soon as TSLint will be integrated into ESLint
 */
export const eslintValidation = async (cwd: string, analytics: ESLintAnalyzerResult): Promise<ValidationResult[]> => {
	if (analytics.eslint) {
		// TODO: eslint is available, so we need to validate all things
		let validationResults: ValidationResult[] = [];
		validationResults = validationResults.concat(
			validateInstallation({
				installation: analytics.eslintInstallation,
				name: 'ESLint',
				source: 'eslintValidation'
			})
		);

		if (!analytics.eslintIgnore) {
			validationResults.push({
				level: ValidationSeverityLevel.warning,
				message: 'No .eslintignore found',
				source: 'eslintValidation'
			})
		}

		return validationResults;
	}

	return [{
		level: ValidationSeverityLevel.warning,
		message: 'ESLint configuration missing',
		source: 'eslintValidation'
	}];
}