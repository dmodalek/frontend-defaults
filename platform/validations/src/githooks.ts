import { GitHooksAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import { ValidationResult, ValidationSeverityLevel, validateInstallation } from '@namics/frontend-defaults-platform-core';

export const gitHooksValidation = async (cwd: string, analytics: GitHooksAnalyzerResult): Promise<ValidationResult[]> => {
	if (analytics.gitHooks) {
		let validationResults: ValidationResult[] = [];
		validationResults = validationResults.concat(
			validateInstallation({
				installation: analytics.gitHooksInstallation,
				name: 'GitHooks',
				source: 'gitHooksValidation'
			})
		);

		if (analytics.gitHooksLocation === 'rc') {
			validationResults.push({
				level: ValidationSeverityLevel.warning,
				message: 'For future projects we decided to use the package.json for configuration (.rc file used)',
				source: 'gitHooksValidation'
			})
		}

		return validationResults;
	}

	return [{
		level: ValidationSeverityLevel.info,
		message: 'GitHooks are missing',
		source: 'gitHooksValidation'
	}];
}