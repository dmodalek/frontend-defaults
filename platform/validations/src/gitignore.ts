import { GitIgnoreAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import { ValidationResult, ValidationSeverityLevel } from '@namics/frontend-defaults-platform-core';

export const gitIgnoreValidation = async (
	cwd: string,
	analytics: GitIgnoreAnalyzerResult
): Promise<ValidationResult[]> => {
	if (analytics.gitignore) {
		// TODO: gitignore is available, so we need to validate the contents
		return [];
	}

	return [
		{
			level: ValidationSeverityLevel.error,
			message: 'Git ignore missing',
			source: 'gitIgnoreValidation',
		},
	];
};
