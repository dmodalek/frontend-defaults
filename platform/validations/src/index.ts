export * from './editorconfig';
export * from './eslint';
export * from './githooks';
export * from './gitignore';
export * from './license';
export * from './nitro';
export * from './node';
export * from '../../patches/src/npmrc';
export * from './readme';
export * from './tslint';
export * from './typescript';
export * from './webpack';
