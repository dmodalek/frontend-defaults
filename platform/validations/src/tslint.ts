import { TSLintAnalyzerResult, TypeScriptAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import {
	ValidationResult,
	ValidationSeverityLevel,
	validateInstallation,
} from '@namics/frontend-defaults-platform-core';

export const tslintValidation = async (
	cwd: string,
	analytics: TypeScriptAnalyzerResult & TSLintAnalyzerResult
): Promise<ValidationResult[]> => {
	if (analytics.tslint) {
		let validationResults: ValidationResult[] = [];
		validationResults = validationResults.concat(
			validateInstallation({
				installation: analytics.tslintInstallation,
				name: 'TSLint',
				source: 'tslintValidation',
			})
		);

		return validationResults;
	}

	// only add issue when typescript is used
	return analytics.typescript
		? [
			{
				level: ValidationSeverityLevel.error,
				message: 'TSLint configuration missing',
				source: 'tslintValidation',
			},
		]
		: [];
};
