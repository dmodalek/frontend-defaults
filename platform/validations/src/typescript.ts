import { TypeScriptAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import {
	checkDependencyUpdate,
	ValidationResult,
	ValidationSeverityLevel,
} from '@namics/frontend-defaults-platform-core';

export const typescriptValidation = async (cwd: string, analytics: TypeScriptAnalyzerResult): Promise<ValidationResult[]> => {
	if (!analytics.typescript) {
		// no typescript found, skip validation
		return [];
	}

	const validations: ValidationResult[] = [];

	if (analytics.typescriptInstallation && !analytics.typescriptInstallation.installed) {
		// TypeScript is not installed, which is not allowed
		validations.push({
			message: `Using TypeScript without explicit installtion is not allowed`,
			source: 'TypeScriptValidation',
			level: ValidationSeverityLevel.error,
			patch: ['TypeScriptInstallationPatch'],
		});
	} else {
		const update = await checkDependencyUpdate(cwd, 'typescript');
		if (update.upgradable) {
			// TypeScript is outdated, upgrade would be an option
			validations.push({
				message: `Please update TypeScript to v${update.latest} (installed: v${update.current})`,
				source: 'TypeScriptValidation',
				level: ValidationSeverityLevel.info,
				patch: [
					['TypeScriptUpdatePatch', {
						current: update.current,
						latest: update.latest
					}]
				],
			});
		}
	}

	return validations;
}