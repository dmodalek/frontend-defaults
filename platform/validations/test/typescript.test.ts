import { ProjectValidator, ProjectAnalyzer, ValidationExceptionLevel } from '@namics/frontend-defaults-platform-core';
import { TypeScriptValidation } from '../src/typescript';
import { getFixtureDir, checkInternetConnection, getLatestVersion } from './utils';
import { TypeScriptAnalyzer, TypeScriptAnalyzerResult } from '../../analyzers/src/typescript';
import { TSLintAnalyzer, TSLintAnalyzerResult } from '../../analyzers/src/tslint';

const getAnalyzerFor = async (
	fixtureContext: string
): Promise<ProjectAnalyzer<TypeScriptAnalyzerResult & TSLintAnalyzerResult>> => {
	return new ProjectAnalyzer<TypeScriptAnalyzerResult & TSLintAnalyzerResult>({
		context: fixtureContext,
		analyzers: [TypeScriptAnalyzer, TSLintAnalyzer],
	}).boot();
};

beforeAll(async () => {
	if (!await checkInternetConnection()) {
		console.debug(`Please only run tests while connected to the internet!`);
		process.exit(1); // kill child worker
	}
});

describe('Validations', () => {
	describe('TypeScriptValidation', () => {
		it('should be creatable without errors', () => {
			expect(async () => {
				const analyzer = await getAnalyzerFor(getFixtureDir('ts-project'));
				new ProjectValidator({
					analyzer,
					validations: [TypeScriptValidation],
				});
			}).not.toThrow();
		});

		it('should validate a project with upgrades correctly', async () => {
			const analyzer = await getAnalyzerFor(getFixtureDir('ts-project'));
			const validator = await new ProjectValidator({
				analyzer,
				context: getFixtureDir('ts-project'),
				validations: [TypeScriptValidation],
			});

			const patches = await validator.validate();
			expect(patches).toHaveLength(1);
			expect(validator.validation).toHaveLength(1);

			const [upgradeInfo] = validator.validation;
			expect(upgradeInfo.level).toEqual(ValidationExceptionLevel.info);
			expect(upgradeInfo.patches).toBeTruthy();
			expect(upgradeInfo.patches![0].patch).toEqual('TypeScriptUpdatePatch');
			expect(upgradeInfo.patches![0].arguments.current).toEqual('3.1.6');
			expect(upgradeInfo.patches![0].arguments.latest).toEqual(await getLatestVersion('typescript'));
			expect(upgradeInfo.toString()).toMatchInlineSnapshot(
				`"Validate TypeScriptValidation: Please update TypeScript to v3.2.4 (installed: v3.1.6) [INFO]"`
			);
		});

		it('should analyze a project without installation correctly', async () => {
			const analyzer = await getAnalyzerFor(getFixtureDir('ts-project-not-installed'));
			const validator = await new ProjectValidator({
				analyzer,
				context: getFixtureDir('ts-project-not-installed'),
				validations: [TypeScriptValidation],
			});

			const patches = await validator.validate();
			expect(patches).toHaveLength(1);
			expect(patches[0].patch).toEqual('TypeScriptInstallationPatch');
			expect(validator.validation[0].toString()).toMatchInlineSnapshot(
				`"Validate TypeScriptValidation: Using TypeScript without explicit installtion is not allowed [ERROR]"`
			);
		});
	});
});
