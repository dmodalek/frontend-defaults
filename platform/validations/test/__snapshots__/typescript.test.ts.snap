// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validations TypeScriptValidation should analyze a project without installation correctly 2`] = `
Array [
  Object {
    "level": "ERROR",
    "message": "Using TypeScript without explicit installtion is not allowed",
    "patch": Array [
      "TypeScriptInstallationPatch",
    ],
    "source": "TypeScriptValidation",
  },
]
`;

exports[`Validations TypeScriptValidation should validate a project with upgrades correctly 2`] = `
Array [
  Object {
    "level": "INFO",
    "message": "Please update TypeScript to v3.3.3 (installed: v3.0.0)",
    "patch": Array [
      Array [
        "TypeScriptUpdatePatch",
        Object {
          "current": "3.0.0",
          "latest": "3.3.3",
        },
      ],
    ],
    "source": "TypeScriptValidation",
  },
]
`;
